_drafts/ota-delta-updates.md

```diff
@@ -0,0 +1,600 @@
+original Jojodiff or Jan's JS version, `jdiff`[^jdiff], which can be installed
+via NPM (`npm install -g jdiff`). I chose the latter.
+
+Jdiff's syntax is simple:
+
+```
+$ jdiff <base-image> <new-image> <output-patch-bin>
+```
+
+Which translates to the following `make` commands:
+
+```
+$(BUILD_DIR)/patch.bin: $(BUILD_DIR)/$(PROJECT)-app.bin $(BUILD_DIR)/$(PROJECT)-app-patched.bin
+	$(JDIFF) $^ $@ &>/dev/null # prints useless warning from emscripten, so ignore it
+```
+
+The resulting `patch.bin` is 1252 bytes, down from 7908 bytes for the full app
+image. This is only a ~6x reduction in size, but in practice you will see much
+better result on larger and more complex applications.
+
+Since we're not concerned with downloading the image from the cloud for the sake
+of this blog post, we just hardcode the patch.bin inside of our Loader image
+using `objcopy` as we explained in previous posts.
+
+> Note: C23 which was just finalized will include a new `#embed` preprocessor
+> directive which makes embedded binary and other files inside of a C program
+> much easier. We'll use that instead of objcopy in the future!
+
+
+### Implementing JanPatch
+
+Janpatch is distributed as a single header file which makes it easy to integrate
+into your firmware build. Simply clone the repository, stick it on your include
+path, and `#include <janpatch.h>` in your C code.
+
+Everything the library needs to work is passed via the `janpatch_ctx` struct.
+Let's take a look at it:
+
+```c
+typedef struct {
+    // fread/fwrite buffers
+    janpatch_buffer source_buffer;
+    janpatch_buffer patch_buffer;
+    janpatch_buffer target_buffer;
+
+    // function signatures
+    size_t (*fread)(void*, size_t, size_t, JANPATCH_STREAM*);
+    size_t (*fwrite)(const void*, size_t, size_t, JANPATCH_STREAM*);
+    int    (*fseek)(JANPATCH_STREAM*, long int, int);
+    long   (*ftell)(JANPATCH_STREAM*);
+
+    // progress callback
+    void   (*progress)(uint8_t);
+
+    // the combination of the size of both the source + patch files (that's the max. the target file can be)
+    long   max_file_size;
+} janpatch_ctx;
+```
+
+First come the source, patch, and target buffers. These are simply RAM buffers
+used by Janpatch as working memory. The author recommends matching the size of
+your flash sectors for best performance, likely because the whole buffer is
+written at once, so you'll save on erase times.
+
+On our STM32, flash sectors are 4kB so we'll simply allocate three static 4kB
+buffers:
+
+```
+#include <janpatch.h>
+
+static unsigned char source_buf[4096];
+static unsigned char target_buf[4096];
+static unsigned char patch_buf[4096];
+
+int do_janpatch() {
+    janpatch_ctx ctx = {
+        // fread/fwrite buffers for every file, minimum size is 1 byte
+        // when you run on an embedded system with block size flash, set it to the size of a block for best performance
+        { source_buf, 4096 },
```

Comment:
Maybe sizeof? Or maybe it's more clear as explicit value

Reply:
Added a `#define`